### ISSUE TITLE
Discussion about client side tampering with options

### ISSUE THREAD

**Issue by lwe** [riapi/riapi#12]
Just stumbled upon riapi via [imageflow.io](https://www.imageflow.io/) and it looks like a really promising spec and really dig to idea to standardize this somewhat.

Not sure if this has been discussed yet (hope I didn't miss any ticket or a paragraph in the spec) - or if this is even considered to be in scope for riapi, but allowing the client to fully control sizing, composing etc. also leaves a door open for bad client behaviour. A bad client I could think of is one that starts to request the following images and filling up the local cache / S3 storage / processing CPU usage:

```
/image.jpg?width=1&height=1
/image.jpg?width=2&height=2
/image.jpg?width=3&height=3
...
```

While not all are probably affected by this - people running a dynamic image resize service (even with a cache in front) on Heroku might be hit by this rather quickly 😄

Personally I think it would maybe make sense to either (A) clearly state the this kind of behaviour needs to be blocked or handled server-side (e.g. rate limiting, filtering allowed options, hiding via a proxy etc.) or (B) provide an option to create paths that can't be tampered with by the client, when they are generated by a trusted party.

One solution for option (B) could be adding `signature=` parameter that is generated on the server side using a shared secret, this is something that probably could work well for most of webapps, blogs and similar use-cases. The basic idea is to generate a signature with the path, params and the shared secret (e.g. using HMAC SHA256) and then appending this to the url, this would allow to be easily checked on the server side and still allows trusted peers to create any image they want, without re-configuring the image server. However this would work not so well I guess for (some) native apps, where each device might request a different image size, depending on the screen size and embedding the shared secret basically voids the safety added. Yet even apps might want to be limited to some fixed sizes...

What are your thoughts and opinions about this topic - or is it out-of-scope for the idea of riapi?

**Comment by nathanaeljones** [comment]
**Definitely not out of scope**

To preface this: I'm not aware of a scenario where generic DDOS protection isn't successful against this kind of attack, nor of a verified instance of it occurring in the wild. We've seen vulnerability scanners randomly vary querystring values but not as an effective or intentional attack, just to detect injection vulnerabilities. 

In general, I advise people to have some kind of network-level attack detection. If you get 10k requests of any kind from a client, they might need to complete a captcha or something. 

----

**Flexible signatures should definitely be part of the spec**. I've done a [lot of early work to see how to permit variations](https://github.com/imazen/resizer/tree/features/requestsigning/Plugins/Security/Authorization) without offering an escape vector, but have published no documentation or spec yet. 

ImageResizer offers 'presets' right now, but this is clumsy since you have to look up what you've defined them to mean. A generic signature system that permits for whitelisted breakpoints would be quite useful. 

Would you be interested in collaborating on this kind of spec?

**Comment by lwe** [comment]
Jup, I'm neither aware of any known attacks. Yes sure, let me know if and how I could help with shaping this part of the spec?

**Comment by nathanaeljones** [comment]
I'm thinking it might be best to start with an implementation or two before formalizing it. What client-side language would you want to focus on for URL generation?

**Comment by lwe** [comment]
Language of choice would be Ruby for me, so something on top of refile would be my proposal (even-though [refile](https://github.com/refile/refile) does not implement the riapi spec).

**Comment by nathanaeljones** [comment]
Cool, I somehow missed refile. It appears to be both a server and client, though?

**Comment by lwe** [comment]
Yes, it provides both, a Rack based server for serving and processing files on the fly and a Rails helpers. Also this is not an image resizer, but more of a generic Rails file upload library, with dynamic processing options.

**Comment by lwe** [comment]
It's an ideal target I'd say, because an implementation needs to tackle both parts, right? Generating and adding a signature on the client and checking / enforcing it on the server.

**Comment by nathanaeljones** [comment]
I'm actually working on a Rust server first, but a ruby client could happen pretty early along. 

@choonkeat might be interested as well for [attache](https://github.com/choonkeat/attache).

**Comment by choonkeat** [comment]
Targeting `refile` specifically is unnecessary... if anything, targeting `rack` would allow smaller players like attache & attache-rails to join riapi easier 🙏

**Comment by choonkeat** [comment]
> Generating and adding a signature on the client and checking / enforcing it on the server.

we actually only need a `generate_signature(params)` method to generate such signature given a params hash (ignores non riapi param keys).

it is a bonus if there's a bundled rack middleware that uses this same method, verify against `params[:signature]` and choose to throw a 401 or pass-through the request

then again, attache has per-hostname secret. so unless this bundled rack middleware is as complicated, i'll need to write my own middleware anyway, using the `generate_signature` method from such a lib
